tests.md

```diff
@@ -0,0 +1,219 @@
+  u_false   = length(random_pet.this.id) < 0
+  u_string  = random_pet.this.id
+  u_number  = length(random_pet.this.id)
+  u_cidr    = "10.0.${local.u_number}.0/24"
+}
+```
+
+This way you can replace a `true` known bool value with `local.u_true` unknown bool value with ease. (The `local.u_true`
+is always equal to `true`, because the `length()` is always greater than zero. But only the Apply stage determines
+that, the Plan stage assumes it can be either true or false.) Same for other types of inputs: you can replace any string
+by `local.u_string`, and number by `local.u_number`, and so on. This way it is possible to set every input of the module
+to an unknown value in quite a readable way.
+
+This pattern uses `random_pet`, a very simple Hashicorp-provided resource which simply generates names,
+e.g. "relevant-horse". It is very fast, does not communicate to any cloud, and does not incur any cost, but otherwise
+does not bring any real functionality here - the code above would be usable even if we would use `aws_vpc` instead.
+
+### Outputs
+
+The majority of module's outputs should be consumable by subsequent Terraform code. Only the minority of outputs
+exist to be merely displayed to a user when Terraform completes. We can test for some very predictable scenarios of how
+our module's output is consumed by some typical user-side code. It turns out that unknown values impact this too.
+
+The main scenario is when an output is a `map`. There is not much one can do with a map, except to perform a series
+of transformations and finally feed it to some `for_each` block at some point. But it's just impossible if the keys
+of the map are an unknown set - the Plan will just fail with "Invalid for_each argument". The module theoretically works
+but prevents any Lego®-style building. (Such a map output can still be displayed to users however.)
+
+```hcl2
+module "mymodule" {
+  source = "../../modules/mymodule"
+
+  /* inputs */
+}
+
+### Feed all the map outputs to a dummy for_each ###
+
+resource "random_pet" "consume_mymodule_my_map_output" {
+  for_each = module.mymodule.my_map_output
+}
+
+resource "random_pet" "consume_mymodule_second" {
+  for_each = module.mymodule.second
+}
+
+resource "random_pet" "consume_mymodule_third" {
+  for_each = module.mymodule.third
+}
+```
+
+On Terraform 0.13 and above condense that to a single block:
+
+```hcl2
+resource "random_pet" "consume_maps" {
+  for_each = merge(
+    module.mymodule.my_map_output,
+    module.mymodule.second,
+    module.mymodule.third,
+
+    module.mymodule_alt.my_map_output,
+    module.mymodule_alt.second,
+    module.mymodule_alt.third,
+  )
+}
+```
+
+The `_alt` above refers to an alternate call to the same module, with a different set of known inputs. For example,
+if the `vpc` module has the input `create_vpc` which is true by default, an alternate call would have it false.
+This would test Plan on multiple execution paths within the same code.
```

Comment:
Am I going overboard? Split it to another doc?

Reply:
IMO - keep it this way, I prefer having this in one bigger doc, than searching through a few smaller ones.

Reply:
👍 Will do so. 👍